src/System.Device.Gpio/System/Device/Gpio/Drivers/LibgpiodDriver.cs

```diff
@@ -0,0 +1,195 @@
+        }
+
+        protected internal override void ClosePin(int pinNumber)
+        {
+            MutableTuple<PinMode, Process> pinModeAndProcess;
+            if (_pinNumberToPinModeAndProcess.TryGetValue(pinNumber, out pinModeAndProcess)) 
+            {
+                Process existingProcess = pinModeAndProcess.Value2;
+                if (existingProcess != null) {
+                    existingProcess.Kill();
+                    existingProcess.WaitForExit(1);
+                    pinModeAndProcess.Value2 = null;
+                }
+                _pinNumberToPinModeAndProcess.Remove(pinNumber);
+            }
+        }
+
+        protected internal override int ConvertPinNumberToLogicalNumberingScheme(int pinNumber) => throw new PlatformNotSupportedException("This driver is generic so it can not perform conversions between pin numbering schemes.");
+
+        protected internal override PinMode GetPinMode(int pinNumber)
+        {
+            MutableTuple<PinMode, Process> pinModeAndProcess;
+            if (_pinNumberToPinModeAndProcess.TryGetValue(pinNumber, out pinModeAndProcess))
+             {
+                return pinModeAndProcess.Value1;
+            }
+            return PinMode.Input;
+        }
+
+        protected internal override bool IsPinModeSupported(int pinNumber, PinMode mode)
+        {
+            // Libgpiod command line tools does not support pull up or pull down resistors.
+            return mode != PinMode.InputPullDown && mode != PinMode.InputPullUp;
+        }
+
+        protected internal override void OpenPin(int pinNumber)
+        {
+            // do nothing
+        }
+
+        protected internal override PinValue Read(int pinNumber)
+        {
+            PinValue result = default(PinValue);
+            Process gpioGetCommand = new Process();
+            gpioGetCommand.StartInfo.FileName = GpioGet;
+            gpioGetCommand.StartInfo.Arguments = $"{GpioChip0} {pinNumber}";
+            gpioGetCommand.StartInfo.RedirectStandardOutput = true;
+            gpioGetCommand.Start();
+            if (WaitForProcessToExit(gpioGetCommand, $"{GpioGet} {GpioChip0} {pinNumber}", TimeoutMilliSeconds))
+            {
+                string valueContents = gpioGetCommand.StandardOutput.ReadToEnd();
+                result = ConvertMachineValueToPinValue(valueContents);
+            }
+            return result;
+        }
+
+        private PinValue ConvertMachineValueToPinValue(string value)
+        {
+            return value != null && value.IndexOf('1') != -1 ? PinValue.High : PinValue.Low;
+        }
+
+        protected internal override void RemoveCallbackForPinValueChangedEvent(int pinNumber, PinChangeEventHandler callback)
+        {
+            throw new NotImplementedException();
+        }
+
+        protected internal override void SetPinMode(int pinNumber, PinMode mode)
+        {
+            if (mode == PinMode.InputPullDown || mode == PinMode.InputPullUp)
```

Comment:
This repeats the code in `IsPinModeSupported`. Can you call it?

Reply:
Well that is exposed method in the interface, i could say controller could call it to avoid an exception